README.md
# fastify-metrics

<div align="center">
  <img src="https://github.com/SkeLLLa/fastify-metrics/raw/master/logo.png" alt="fastify-metrics logo"/>
</div>

[![NPM Version](https://img.shields.io/npm/v/fastify-metrics.svg)](https://www.npmjs.com/package/fastify-metrics)
[![Downloads Count](https://img.shields.io/npm/dm/fastify-metrics.svg)](https://www.npmjs.com/package/fastify-metrics)
[![Vunerabilities Count](https://snyk.io/test/npm/fastify-metrics/badge.svg)](https://www.npmjs.com/package/fastify-metrics)
[![Build Status](https://github.com/SkeLLLa/fastify-metrics/workflows/build/badge.svg)](https://github.com/SkeLLLa/fastify-metrics/actions)
[![License](https://img.shields.io/npm/l/fastify-metrics.svg)](https://gitlab.com/m03geek/fastify-metrics/blob/master/LICENSE)
[![Codecov](https://img.shields.io/codecov/c/gh/SkeLLLa/fastify-metrics.svg)](https://codecov.io/gh/SkeLLLa/fastify-metrics)
[![LGTM Alerts](https://img.shields.io/lgtm/alerts/github/SkeLLLa/fastify-metrics.svg)](https://lgtm.com/projects/g/SkeLLLa/fastify-metrics/)
[![LGTM Grade](https://img.shields.io/lgtm/grade/javascript/github/SkeLLLa/fastify-metrics.svg)](https://lgtm.com/projects/g/SkeLLLa/fastify-metrics/)

[Prometheus](https://prometheus.io/) metrics exporter for Fastify.

This plugin uses [prom-client](https://github.com/siimon/prom-client) under the hood.

This plugin also adds two http metrics for your routes:

- Requests duration histogram
- Requests duration summary

## ToC

- [fastify-metrics](#fastify-metrics)
  - [ToC](#toc)
  - [Fastify support](#fastify-support)
  - [Notable changes](#notable-changes)
    - [v10.x.x](#v10xx)
    - [v9.x.x](#v9xx)
    - [v6.x.x](#v6xx)
  - [Installation](#installation)
  - [Features and requirements](#features-and-requirements)
  - [Usage](#usage)
    - [Registry clear](#registry-clear)
    - [Plugin options](#plugin-options)
      - [Route metrics](#route-metrics)
        - [Route metrics overrides](#route-metrics-overrides)
          - [Labels](#labels)
          - [Request durations summary](#request-durations-summary)
          - [Request durations histogram](#request-durations-histogram)
    - [HTTP routes metrics in Prometheus](#http-routes-metrics-in-prometheus)
  - [API Docs](#api-docs)
  - [Changelog](#changelog)
  - [See also](#see-also)
  - [License](#license)

## Fastify support

- **v3.x.x** - supports `fastify-1.x`
- **v4.x.x** - supports `fastify-2.x` `prom-client-11.x`
- **v5.x.x** - supports `fastify-2.x` `prom-client-12.x`
- **v6.x.x** - supports `fastify-3.x`
- **v9.x.x** - supports `fastify-4.x` `prom-client-14.x`

## Notable changes

### v10.x.x

- Replace route `context.config` with `routeConfig` due to deprecation in fastify v4 and removal in fastify v5. If you had `disableMetrics` option in you route `config`, update fastify to latest version.

### v9.x.x

- Fastify v4 support.
- Complete config rewrite, default behaviour changed.
- Support disabling metrics in route config.
- Now collects metrics only for registered routes by default.
- Unknown routes metrics collection disabled by default.
- Removed `metrics` from `request`. Now it uses `WeakMap` and not exposed.
- Add balcklisting possibility for request methods.
- Registry overrides moved to metric configuration.
- Support overriding all Summary and Histogram options for default route metrics.

### v6.x.x

- Fastify v3 support.
- Drop node.js 8 support.
- `enableDefaultMetrics` - now enables only default `prom-client` metrics. Set to `true` by default.
- `enableRouteMetrics` - additional flag that enables route metrics. Set to `true` by default.

## Installation

```sh
npm i fastify-metrics --save
pnpm i fastify-metrics --save
```

<sub>[Back to top](#toc)</sub>

## Features and requirements

- Collects default server metrics (see [prom-client](https://github.com/siimon/prom-client/tree/master/lib/metrics));
- Collects route response timings
- Adds `metrics` to fastify instance for your custom metrics.

---

- Requires fastify `>=4.0.0`.
- Node.js `>=16.0.0`.

<sub>[Back to top](#toc)</sub>

## Usage

Add it to your project like regular fastify plugin. Use `register` method and pass options to it.

```js
const fastify = require('fastify');
const app = fastify();

const metricsPlugin = require('fastify-metrics');
await app.register(metricsPlugin, { endpoint: '/metrics' });
```

It also exports client to fastify instance `fastify.metrics.client` which you may use it in your routes.

You may create your metrics when app starts and store it in `fastify.metrics` object and reuse them in multiple routes.

### Registry clear

After calling `registry.clear()` all metrics are removed from registry. In order to add them again to the registry, call `fastify.mterics.initMetricsInRegistry`.

<sub>[Back to top](#toc)</sub>

### Plugin options

See for details [docs](docs/fastify-metrics.imetricspluginoptions.md)

| Property                                                                          | Type                                                                                                            | Default Value       |
| --------------------------------------------------------------------------------- | --------------------------------------------------------------------------------------------------------------- | ------------------- |
| [defaultMetrics?](./docs/fastify-metrics.imetricspluginoptions.defaultmetrics.md) | [IDefaultMetricsConfig](./docs/fastify-metrics.idefaultmetricsconfig.md)                                        | `{ enabled: true }` |
| [endpoint?](./docs/fastify-metrics.imetricspluginoptions.endpoint.md)             | string \| null \| [`Fastify.RouteOptions`](https://www.fastify.io/docs/latest/Reference/Routes/#routes-options) | `'/metrics'`        |
| [name?](./docs/fastify-metrics.imetricspluginoptions.name.md)                     | string                                                                                                          | `'metrics'`         |
| [routeMetrics?](./docs/fastify-metrics.imetricspluginoptions.routemetrics.md)     | [IRouteMetricsConfig](./docs/fastify-metrics.iroutemetricsconfig.md)                                            | `{ enabled: true }` |

#### Route metrics

| Property                                                                                    | Type                                                                                          | Default Value                           |
| ------------------------------------------------------------------------------------------- | --------------------------------------------------------------------------------------------- | --------------------------------------- |
| [enabled?](./docs/fastify-metrics.iroutemetricsconfig.enabled.md)                           | boolean                                                                                       | `true`                                  |
| [groupStatusCodes?](./docs/fastify-metrics.iroutemetricsconfig.groupstatuscodes.md)         | boolean                                                                                       | `false`                                 |
| [invalidRouteGroup?](./docs/fastify-metrics.iroutemetricsconfig.invalidroutegroup.md)       | string                                                                                        | `'__unknown__'`                         |
| [methodBlacklist?](./docs/fastify-metrics.iroutemetricsconfig.methodblacklist.md)           | readonly string\[\]                                                                           | `['HEAD','OPTIONS','TRACE','CONNECT',]` |
| [overrides?](./docs/fastify-metrics.iroutemetricsconfig.overrides.md)                       | [IRouteMetricsOverrides](./docs/fastify-metrics.iroutemetricsoverrides.md)                    |                                         |
| [registeredRoutesOnly?](./docs/fastify-metrics.iroutemetricsconfig.registeredroutesonly.md) | boolean                                                                                       | `true`                                  |
| [customLabels?](./fastify-metrics.iroutemetricsconfig.customlabels.md)                      | Record&lt;string, string \| ((request: FastifyRequest, reply: FastifyReply) =&gt; string)&gt; | `undefined`                             |
| [routeBlacklist?](./docs/fastify-metrics.iroutemetricsconfig.routeblacklist.md)             | readonly string\[\]                                                                           | `[]`                                    |

##### Route metrics overrides

You may override default metrics settings. You may provide overrides for two metrics tracking http request durations: `histogram` and `summary`.

```js
const fastify = require('fastify');
const app = fastify();
const metricsPlugin = require('fastify-metrics');

await app.register(metricsPlugin, {
  endpoint: '/metrics',
  routeMetrics: {
    overrides: {
      histogram: {
        name: 'my_custom_http_request_duration_seconds',
        buckets: [0.1, 0.5, 1, 3, 5],
      },
      summary: {
        help: 'custom request duration in seconds summary help',
        labelNames: ['status_code', 'method', 'route'],
        percentiles: [0.5, 0.75, 0.9, 0.95, 0.99],
      },
    },
  },
});
```

###### Labels

| Property                                                                   | Type                                   | Default value   |
| -------------------------------------------------------------------------- | -------------------------------------- | --------------- |
| [getRouteLabel?](./fastify-metrics.iroutelabelsoverrides.getroutelabel.md) | (request: FastifyRequest) =&gt; string | `undefined`     |
| [method?](./docs/fastify-metrics.iroutelabelsoverrides.method.md)          | string                                 | `'method'`      |
| [route?](./docs/fastify-metrics.iroutelabelsoverrides.route.md)            | string                                 | `'route'`       |
| [status?](./docs/fastify-metrics.iroutelabelsoverrides.status.md)          | string                                 | `'status_code'` |

###### Request durations summary

| Property                                                                | Type       | Default value                           |
| ----------------------------------------------------------------------- | ---------- | --------------------------------------- |
| [name?](./docs/fastify-metrics.isummaryoverrides.name.md)               | string     | `'http_request_summary_seconds'`        |
| [help?](./docs/fastify-metrics.isummaryoverrides.help.md)               | string     | `'request duration in seconds summary'` |
| [percentiles?](./docs/fastify-metrics.isummaryoverrides.percentiles.md) | number\[\] | `[0.5, 0.9, 0.95, 0.99]`                |

###### Request durations histogram

| Property                                                          | Type       | Default value                     |
| ----------------------------------------------------------------- | ---------- | --------------------------------- |
| [name?](./docs/fastify-metrics.ihistogramoverrides.name.md)       | string     | `'http_request_duration_seconds'` |
| [help?](./docs/fastify-metrics.ihistogramoverrides.help.md)       | string     | `'request duration in seconds'`   |
| [buckets?](./docs/fastify-metrics.ihistogramoverrides.buckets.md) | number\[\] | `[0.05, 0.1, 0.5, 1, 3, 5, 10]`   |

<sub>[Back to top](#toc)</sub>

### HTTP routes metrics in Prometheus

The following table shows what metrics will be available in Prometheus. Note suffixes like `_bucket`, `_sum`, `_count` are added automatically.

| metric                                 | labels                           | description                   |
| -------------------------------------- | -------------------------------- | ----------------------------- |
| `http_request_duration_seconds_count`  | `method`, `route`, `status_code` | Requests total count          |
| `http_request_duration_seconds_bucket` | `method`, `route`, `status_code` | Requests durations by bucket  |
| `http_request_summary_seconds`         | `method`, `route`, `status_code` | Requests duration percentiles |
| `http_request_summary_seconds_count`   | `method`, `route`, `status_code` | Requests total count          |

<sub>[Back to top](#toc)</sub>

## API Docs

See [docs](docs/index.md).

<sub>[Back to top](#toc)</sub>

## Changelog

See [changelog](CHANGELOG.md).

<sub>[Back to top](#toc)</sub>

## See also

- [fastify-prom-client](https://github.com/ExcitableAardvark/fastify-prom-client) - just simple client that adds aggregated http requests metric.

<sub>[Back to top](#toc)</sub>

## License

Licensed under [MIT](./LICENSE).

<sub>[Back to top](#toc)</sub>


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## [10.2.0](https://github.com/SkeLLLa/fastify-metrics/compare/v10.0.4...v10.2.0) (2023-03-16)


### Features

* configurable route label value override ([01134f9](https://github.com/SkeLLLa/fastify-metrics/commit/01134f9cbaf4fe18440f0ef99d48851434b5c0f1))
* custom labels ([1f30671](https://github.com/SkeLLLa/fastify-metrics/commit/1f30671991b03ed1b9730934bf78d1480d4e842d))
* tests for custom labels ([d0a8422](https://github.com/SkeLLLa/fastify-metrics/commit/d0a84224971988e1a473177eb5560c3d0cd5595b))
* update docs ([6ccaade](https://github.com/SkeLLLa/fastify-metrics/commit/6ccaadeee53ef4e189a04f7cfe61106720432532))


### Misc

* **release:** 10.1.0 ([1d8f06a](https://github.com/SkeLLLa/fastify-metrics/commit/1d8f06ac5ea8f9cff3dc70841f08761229cbe98d))

## [10.1.0](https://github.com/SkeLLLa/fastify-metrics/compare/v10.0.4...v10.1.0) (2023-03-14)


### Features

* configurable route label value override ([01134f9](https://github.com/SkeLLLa/fastify-metrics/commit/01134f9cbaf4fe18440f0ef99d48851434b5c0f1))
* update docs ([6ccaade](https://github.com/SkeLLLa/fastify-metrics/commit/6ccaadeee53ef4e189a04f7cfe61106720432532))

### [10.0.4](https://github.com/SkeLLLa/fastify-metrics/compare/v10.0.3...v10.0.4) (2023-03-12)

### [10.0.3](https://github.com/SkeLLLa/fastify-metrics/compare/v10.0.2...v10.0.3) (2023-02-07)


### Bug fixes

* fix routes with array of method ([07ab256](https://github.com/SkeLLLa/fastify-metrics/commit/07ab2567c81c21071b4aa81cbd5a98babd896f23))

### [10.0.2](https://github.com/SkeLLLa/fastify-metrics/compare/v10.0.1...v10.0.2) (2023-01-05)

### [10.0.1](https://github.com/SkeLLLa/fastify-metrics/compare/v10.0.0...v10.0.1) (2022-12-27)


### Bug fixes

* fixes typo in README.md ([ae72c72](https://github.com/SkeLLLa/fastify-metrics/commit/ae72c7218e633419ad9feeee5e34ac116b96a9e6))

## [10.0.0](https://github.com/SkeLLLa/fastify-metrics/compare/v9.2.4...v10.0.0) (2022-10-18)


### ⚠ BREAKING CHANGES

* replace context config with routeConfig

### Features

* add notes for v10 ([46beccf](https://github.com/SkeLLLa/fastify-metrics/commit/46beccffe09d0e5456e7813d23927d13df421c13))


### Misc

* **deps:** update dependencies ([ca7efa5](https://github.com/SkeLLLa/fastify-metrics/commit/ca7efa5f3607997d12599bc122d41fccb878e8b8))
* **deps:** update lockfile ([25dbda9](https://github.com/SkeLLLa/fastify-metrics/commit/25dbda973351fc46c5159dcfb74d5962f2b398bf))


### Codestyle

* **lint:** fix lint ([9fee64e](https://github.com/SkeLLLa/fastify-metrics/commit/9fee64edd985b3326fae339a538930bb530386ac))

### [9.2.4](https://github.com/SkeLLLa/fastify-metrics/compare/v9.2.3...v9.2.4) (2022-10-17)

### [9.2.3](https://github.com/SkeLLLa/fastify-metrics/compare/v9.2.2...v9.2.3) (2022-10-17)

### [9.2.2](https://github.com/SkeLLLa/fastify-metrics/compare/v9.2.1...v9.2.2) (2022-08-28)


### Bug fixes

* add plugin name for dependencies ([ec7b04a](https://github.com/SkeLLLa/fastify-metrics/commit/ec7b04ac752fa6a724991dba966683063f2c999e))


### Misc

* **deps:** update ([78c86ed](https://github.com/SkeLLLa/fastify-metrics/commit/78c86edf403f33f86ecb5ddae9954b3ff0b147d3))


### Codestyle

* lint fix ([2a429cd](https://github.com/SkeLLLa/fastify-metrics/commit/2a429cd13d7fe254cef2d61f90002c4084d0815d))

### [9.2.1](https://github.com/SkeLLLa/fastify-metrics/compare/v9.2.0...v9.2.1) (2022-07-28)


### Bug fixes

* **registries:** add workaround to save default labels ([51247ce](https://github.com/SkeLLLa/fastify-metrics/commit/51247cec3b2167e820d799d669a77961f74bac5c))


### Misc

* **deps:** update ([68d80a7](https://github.com/SkeLLLa/fastify-metrics/commit/68d80a72809c45969445e5ac92a9530c457feef2))

## [9.2.0](https://github.com/SkeLLLa/fastify-metrics/compare/v9.1.3...v9.2.0) (2022-07-26)


### Features

* support endpoint option as object ([a993277](https://github.com/SkeLLLa/fastify-metrics/commit/a99327723a51b015c64859d18bb42dd2c3e266ba))

### [9.1.3](https://github.com/SkeLLLa/fastify-metrics/compare/v9.1.2...v9.1.3) (2022-07-21)

### [9.1.2](https://github.com/SkeLLLa/fastify-metrics/compare/v9.1.1...v9.1.2) (2022-07-21)

### [9.1.1](https://github.com/SkeLLLa/fastify-metrics/compare/v9.1.0...v9.1.1) (2022-07-20)

### Misc

- rebuild version ([d22752e](https://github.com/SkeLLLa/fastify-metrics/commit/d22752e7cf369b220ceaf0e30da781f28fede817))

## [9.1.0](https://github.com/SkeLLLa/fastify-metrics/compare/v9.0.2...v9.1.0) (2022-07-20)

### Features

- add initMetricsInRegistry function ([e37a028](https://github.com/SkeLLLa/fastify-metrics/commit/e37a0286297b5b0a4e5b0fec58dba186f4e158b8))

### Bug fixes

- **metrics:** add initMetricsInRegistry method ([dafbb59](https://github.com/SkeLLLa/fastify-metrics/commit/dafbb5921c18900a16f655297f5f03169eb8ef2f))

### Misc

- **release:** 9.0.3 ([56ae239](https://github.com/SkeLLLa/fastify-metrics/commit/56ae239ce6e646b34c21b3eacea931700bf37e36))

### [9.0.3](https://github.com/SkeLLLa/fastify-metrics/compare/v9.0.2...v9.0.3) (2022-07-20)

### Bug fixes

- **metrics:** add initMetricsInRegistry method ([dafbb59](https://github.com/SkeLLLa/fastify-metrics/commit/dafbb5921c18900a16f655297f5f03169eb8ef2f))

### [9.0.2](https://github.com/SkeLLLa/fastify-metrics/compare/v9.0.1...v9.0.2) (2022-07-20)

### [9.0.1](https://github.com/SkeLLLa/fastify-metrics/compare/v9.0.0...v9.0.1) (2022-07-20)

### Misc

- **deps:** update dependencies ([a50d379](https://github.com/SkeLLLa/fastify-metrics/commit/a50d37933f6de3213de3fc2deb43248375ff8993))

## [9.0.0](https://github.com/SkeLLLa/fastify-metrics/compare/v8.0.0...v9.0.0) (2022-07-03)

### ⚠ BREAKING CHANGES

- **upgrade:** new configuration (incompatible with all previous versions)

### Features


lib/index.js
const fp = require('fastify-plugin');
const client = require('prom-client');

const fastifyMetrics = (
  fastify,
  {
    enableDefaultMetrics = true,
    groupStatusCodes = false,
    pluginName = 'metrics',
    interval = 5000,
    blacklist,
    register,
    prefix,
    endpoint,
    metrics = {},
  } = {},
  next
) => {
  const plugin = {client};

  if (typeof blacklist === 'string') {
    blacklist = new RegExp(blacklist, 'i');
  }

  const collectMetricsForUrl = (url) => {
    const queryIndex = url.indexOf('?');
    url = queryIndex === -1 ? url : url.substring(0, queryIndex);
    if (!blacklist) {
      return true;
    }
    if (Array.isArray(blacklist)) {
      return blacklist.indexOf(url) === -1;
    }
    if (typeof blacklist.test === 'function') {
      return !blacklist.test(url);
    }
    return blacklist !== url;
  };

  if (enableDefaultMetrics) {
    const defaultOpts = {timeout: interval};
    const opts = {
      histogram: {
        name: 'http_request_duration_seconds',
        help: 'request duration in seconds',
        labelNames: ['status_code', 'method', 'route'],
        buckets: [0.005, 0.05, 0.1, 0.5, 1, 3, 5, 10],
      },
      summary: {
        name: 'http_request_summary_seconds',
        help: 'request duration in seconds summary',
        labelNames: ['status_code', 'method', 'route'],
        percentiles: [0.5, 0.9, 0.95, 0.99],
      },
    };

    if (register) {
      if (Array.isArray(register)) {
        defaultOpts.register = register[0];
        opts.histogram.registers = register;
        opts.summary.registers = register;
      } else {
        defaultOpts.register = register;
        opts.histogram.registers = [register];
        opts.summary.registers = [register];
      }
    }

    if (prefix) {
      defaultOpts.prefix = prefix;
      opts.histogram.name = `${prefix}${opts.count.name}`;
      opts.summary.name = `${prefix}${opts.count.name}`;
    }

    Object.keys(metrics)
      .filter(opts.hasOwnProperty)
      .forEach((key) => {
        Object.assign(opts[key], metrics[key]);
      });

    client.collectDefaultMetrics(defaultOpts);
    const routeHist = new client.Histogram(opts.histogram);
    const routeSum = new client.Summary(opts.summary);

    if (endpoint) {
      fastify.route({
        url: endpoint,
        method: 'GET',
        schema: {hide: true},
        handler: (_, reply) => {
          const data = register
            ? register.metrics()
            : client.register.metrics();
          reply.type('text/plain').send(data);
        },
      });
    }

    fastify.addHook('onRequest', (request, _, next) => {
      if (collectMetricsForUrl(request.url)) {
        request.metrics = {
          hist: routeHist.startTimer(),
          sum: routeSum.startTimer(),
        };
      }
      next();
    });

    fastify.addHook('onSend', function(request, reply, _, next) {
      if (request.raw.metrics) {
        let routeId = reply.context.config.url || request.raw.url;
        if (reply.context.config.statsId) {
          routeId = reply.context.config.statsId;
        }
        const method = request.raw.method;
        const statusCode = groupStatusCodes
          ? `${Math.floor(reply.res.statusCode / 100)}xx`
          : reply.res.statusCode;

        request.raw.metrics.sum({
          method,
          route: routeId,
          status_code: statusCode,
        });
        request.raw.metrics.hist({
          method,
          route: routeId,
          status_code: statusCode,
        });
      }
      next();
    });
  }
  fastify.decorate(pluginName, plugin);
  next();
};

module.exports = fp(fastifyMetrics, {
  fastify: '>=1.9.0',
  name: 'fastify-metrics',
});
